### PR TITLE
SmoothQuant mappings tutorial

### PR DESCRIPTION
**Pull request by rahul-tuli**
### Description:

This PR adds a tutorial that guides users on specifying the correct mappings for the **SmoothQuant Modifier** in `llm-compressor`. The tutorial explains how to smooth inputs to the `q/k/v` projections in self-attention and the `fc1` block in feed-forward layers, based on the [SmoothQuant paper](https://arxiv.org/abs/2211.10438).

### Changes:

- **New Tutorial**:
  - Located at `src/llmcompressor/modifiers/smoothquant/README.md` the tutorial provides instructions for identifying layers, targeting leaf modules, and using regular expressions for mappings.
  - A sample mapping for LLaMA-like models is included.

### Motivation:

This tutorial simplifies applying SmoothQuant, helping users correctly target layers for efficient model quantization.

### Testing:

- Manually reviewed and tested on a sample LLaMA model to ensure accuracy of instructions.

### Relevant Links:

- SmoothQuant Paper: https://arxiv.org/abs/2211.10438
- Code Reference: [`smoothquant/base.py`](https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/modifiers/smoothquant/base.py)

### CODE REVIEW

File: src/llmcompressor/modifiers/smoothquant/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+A mapping in SmoothQuant takes the form:
+
+```python
+([layers smoothed inputs pass into], output_to_smooth)
+```
+
+For example, in the default mapping:
+```python
+[["re:.*gate_proj", "re:.*up_proj"], "re:.*post_attention_layernorm"]
```
Comment (dsikka) [inline src/llmcompressor/modifiers/smoothquant/README.md:42]: Is one supposed to be square and one on line 37 circle brackets?
I dont really follow what "layers smoothed inputs pass into" means

File: src/llmcompressor/modifiers/smoothquant/README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+```
+This specifies that we want to smooth the inputs feeding into the projections (`gate_proj`, `up_proj`) as well as the output from `post_attention_layernorm`.
+
+## Specifying Your Own Mappings
+
+To create your own mappings, follow these steps:
+
+1. **Identify the layers you want to smooth inputs into**:
```
Comment (dsikka) [inline src/llmcompressor/modifiers/smoothquant/README.md:50]: maybe `to pass smooth inputs`?

File: src/llmcompressor/modifiers/smoothquant/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+2. **Match leaf modules**:
+   Ensure you're targeting leaf modules (i.e., the individual components of broader blocks, such as `gate_proj` and `up_proj` instead of a larger `mlp` module).
+
+3. **Specify the correct regular expressions**:
+   Use regular expressions to match the layers you want to target. For instance, if you want to target all projection layers across all attention heads, you could use a regex like `"re:.*proj"`. If you want to target a specific projection layer, make the regex more specific.
+
+### Example Custom Mapping
+
+Let’s say you’re working with a model that has layers named similarly to LLaMA, and you want to smooth the input activations of the self-attention layers as well as the feed-forward layers. Here’s how you might specify the mapping:
```
Comment (dsikka) [inline src/llmcompressor/modifiers/smoothquant/README.md:61]: `Here is`

**Review by dsikka — COMMENTED** [review]
Followed for the most part apart from one comment for clarity. Maybe throw this into a grammar editor to help with some of the wording. Otherwise, LGTM

File: src/llmcompressor/modifiers/smoothquant/README.md (hunk 1 above)
Comment (rahul-tuli) [reply src/llmcompressor/modifiers/smoothquant/README.md:42]: Both would work in code, but updated the one on line 37 for consistency

File: src/llmcompressor/modifiers/smoothquant/README.md (hunk 2 above)
Comment (rahul-tuli) [reply src/llmcompressor/modifiers/smoothquant/README.md:50]: updated to pass smoothed input activations

File: src/llmcompressor/modifiers/smoothquant/README.md (hunk 1 above)
Comment (dsikka) [reply src/llmcompressor/modifiers/smoothquant/README.md:42]: What does `layers smoothed inputs pass into` mean?

**Review by dsikka — COMMENTED** [review]
just a clarity question otherwise LGTM.

@robertgshaw2-neuralmagic would be good to get your feedback

**Comment by rahul-tuli** [comment]
> just a clarity question otherwise LGTM.
> 
> @robertgshaw2-neuralmagic would be good to get your feedback

Dipika smoothquant shifts the difficulty of quantizing activations to weights by scaling up the the weights and performing the inverse transformation to activations; to accomplish this it needs to know which layer the smoothed out activations pass into i.e which layers weights need to be scaled up.

**Comment by dsikka** [comment]
> > just a clarity question otherwise LGTM.
> > @robertgshaw2-neuralmagic would be good to get your feedback
> 
> Dipika smoothquant shifts the difficulty of quantizing activations to weights by scaling up the the weights and performing the inverse transformation to activations; to accomplish this it needs to know which layer the smoothed out activations pass into i.e which layers weights need to be scaled up.

Yes I understand that.  My comment was that I did not understand the wording behind `layers smoothed inputs pass into`  - read awkwardly. Maybe something like `layers that have smoothed inputs being passed into` or something along those lines

**Comment by rahul-tuli** [comment]
> > > just a clarity question otherwise LGTM.
> > > @robertgshaw2-neuralmagic would be good to get your feedback
> > 
> > 
> > Dipika smoothquant shifts the difficulty of quantizing activations to weights by scaling up the the weights and performing the inverse transformation to activations; to accomplish this it needs to know which layer the smoothed out activations pass into i.e which layers weights need to be scaled up.
> 
> Yes I understand that. My comment was that I did not understand the wording behind `layers smoothed inputs pass into` - read awkwardly. Maybe something like `layers that have smoothed inputs being passed into` or something along those lines

I tried to address it in my last commit kindly let me know if that's better

File: src/llmcompressor/modifiers/smoothquant/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+This means that we need to smooth the inputs feeding into:
+- The **q/k/v blocks** (query, key, value blocks of self-attention)
+- The **fc1 block** (the fully connected block of the feed-forward layer)
+
+We can derive this by examining the diagram on page 5 of the SmoothQuant paper. It shows that smoothing should occur at specific points in the neural network architecture.
+
+### Layer Selection
+
+To get the correct input for smoothing:
+1. For **q/k/v blocks**, we need to target the outputs of `input_layernorm`, as these provide the inputs for the self-attention mechanism.
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/smoothquant/README.md:19]: Surely this isn't guaranteed for all transformer models, right?

File: src/llmcompressor/modifiers/smoothquant/README.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+2. For the **fc1 block**, we need to target the outputs of `post_attention_layernorm`.
+
+### Why Target Leaf Modules?
+
+One of the quirks of working with LLM architectures is that we need to apply smoothing at the leaf nodes of the computational graph. This is why we see mappings such as:
```
Comment (kylesayrs) [inline src/llmcompressor/modifiers/smoothquant/README.md:24]: This doesn't really explain *why* we target leaf modules, only that "it's a quirk". Seems like our software is set up to target leaf modules of any pytorch module, nothing about it is specific to the transformer architecture

**Comment by kylesayrs** [comment]
Why do inputs *and* outputs need to be specified in the mappings? Correct me if I'm wrong, but smoothquant is non-applicable for layers with multiple inputs, so specifying the input module should be enough to uniquely identify that activation, right?

File: src/llmcompressor/modifiers/smoothquant/README.md (hunk 5 above)
Comment (rahul-tuli) [reply src/llmcompressor/modifiers/smoothquant/README.md:24]: Again this is based on the paper, so updated the language to say so!

**Comment by rahul-tuli** [comment]
> Why do inputs _and_ outputs need to be specified in the mappings? Correct me if I'm wrong, but smoothquant is non-applicable for layers with multiple inputs, so specifying the input module should be enough to uniquely identify that activation, right?

You are correct we could do something like that, but that's an algorithmic change which is out of the scope for this PR

**Comment by kylesayrs** [comment]
@rahul-tuli We can make a future change to support output inference, conditioned on the type of the mapping. Ie `List[List[Union[List[str], str]]` is the current mapping type, `List[str]` can be an inferrable mapping type.

**Review by kylesayrs — APPROVED** [review]

**Review by mgoin — APPROVED** [review]

**Pull request merged**